pennylane_lightning/core/src/utils/NDPermuter.hpp

```diff
@@ -0,0 +1,268 @@
+                             const std::vector<std::string> &current_order,
+                             const std::vector<std::string> &new_order) {
+        const std::set<std::string> idx_old(current_order.begin(),
+                                            current_order.end());
+        const std::set<std::string> idx_new(new_order.begin(), new_order.end());
+        const auto data_size = std::accumulate(shape.begin(), shape.end(), 1,
+                                               std::multiplies<std::size_t>());
+        PL_ABORT_IF_NOT(idx_old.size() == current_order.size(),
+                        "Duplicate existing indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(idx_new.size() == new_order.size(),
+                        "Duplicate transpose indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(shape.size() == new_order.size(),
+                        "Tensor shape does not match number of indices.");
+        PL_ABORT_IF_NOT(data_size == data_in.size(),
+                        "Tensor shape does not match given tensor data.");
+        PL_ABORT_IF_NOT(
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
+
+        PL_ABORT_IF(shape.empty(), "Tensor shape cannot be empty.");
+        PL_ABORT_IF(new_order.empty(), "Tensor indices cannot be empty.");
+        return permuter_b_.Transpose(data_in, shape, current_order, new_order);
+    }
+
+  protected:
+    friend PermuterBackend;
+
+  private:
+    PermuterBackend permuter_b_;
+};
+
+/**
+ * @brief Default Permuter backend class for generalised transforms. Adapted
+ * from QFlex.
+ *
+ * @tparam blocksize Controls the internal data chunk size for cache blocking.
+ */
+template <size_t BLOCKSIZE = 1024> class DefaultPermuter {
+
+  public:
+    /**
+     * @brief Reference-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    void Transpose(const std::vector<T> &data_,
+                   const std::vector<size_t> &shape, std::vector<T> &data_out,
+                   const std::vector<std::string> &old_indices,
+                   const std::vector<std::string> &new_indices) {
+        data_out = data_;
+
+        if (new_indices == old_indices)
+            return;
+
+        const std::size_t num_indices = old_indices.size();
+        const std::size_t total_dim = data_.size();
+        std::size_t remaining_data = total_dim;
+
+        if (num_indices == 0) {
+            PL_ABORT("Number of indices cannot be zero.");
+        }
+
+        // Create map_old_to_new_idxpos from old to new indices, and
+        // new_dimensions.
+        std::vector<size_t> map_old_to_new_idxpos(num_indices);
+        std::vector<size_t> new_dimensions(num_indices);
+        for (size_t i = 0; i < num_indices; ++i) {
```

Comment:
would you like to use OMP here?

Reply:
I think we should avoid any optimisation of the direct-from-JET functions now. If it is working, we can remove it for the multi-threading version later, which requires more integration work.

Reply:
To be resolved by later importing directly from JET. Resolving.